pkg/applicationserver/applicationserver.go

```diff
@@ -1055,3 +1065,33 @@ func (as *ApplicationServer) GetMQTTConfig(ctx context.Context) (*config.MQTT, e
 	}
 	return &config.MQTT, nil
 }
+
+func (as *ApplicationServer) getCachedEndDevice(ctx context.Context, ids ttnpb.EndDeviceIdentifiers) (*ttnpb.EndDevice, error) {
+	if as.deviceCache != nil {
+		dev, err := as.deviceCache.Get(ctx, ids)
+		if err != nil {
+			return dev, nil
+		}
+	}
+	cc, err := as.GetPeerConn(ctx, ttnpb.ClusterRole_ENTITY_REGISTRY, ids)
+	if err != nil {
+		return nil, err
+	}
+	client := ttnpb.NewEndDeviceRegistryClient(cc)
+	dev, err := client.Get(ctx, &ttnpb.GetEndDeviceRequest{
+		EndDeviceIdentifiers: ids,
+		FieldMask: pbtypes.FieldMask{
+			Paths: []string{
+				"locations",
```

Comment:
This is lots of magic; how does the caller know that only `locations` are fetched? That happens to be the first use case of getting the device from IS. But this should be passed as method argument.